[WIP] New package: boot-66serv-2.4.1.

mobinmob opened the pull request:
This PR is the continuation of the packaging and integration work @teldra has done on https://github.com/void-linux/void-packages/pull/23122, which is based on the work @zenfailure has done on https://github.com/void-linux/void-packages/pull/21142. They have  done pretty much all the basic work needed -and then some- to integrate @Obarun 's work with void's packaging.

In this PR I will continue their work, try to make the configuration and switch from one init to the other as smooth as possible and package the frontend service files from https://github.com/mobinmob/void-66-services once they are ready.

mobinmob commented:
Version 2.2.1 is a minor update with nice documentation fixes.

mobinmob commented:
Improvements in the runit service which not does not execute /etc/runit/2 anymore:

- The service uses now its own dedicated runlevel  "66" and not the "default" runlevel, which has services that are handled by the boot@ module service (aggety-*).
- It does not try to execute /etc/rc.local if present - that is handled by the services in the boot@ module.
- It does not try to match a word in the kernel commandline to a runlevel and use it.

The above make easier to switch between 66 and runit as the collections of services are seperate. It sould be totaly transparrent for the new user. Existing users that have the runit service enabled should reenable it by:

```
66-disable -S runit
66-enable -S -F -C runit
```
 in order to use the new service frontend. They will also have to reenable the runit services they use the normal way - by symlinking to /var/service.

mobinmob commented:
`@execute` in [stop] is written now in shell and works as it should.

mobinmob commented:
- The switch-initutils service frontend file now resides  in this package instead of the void-66-services repo. The service is needed regardless of the use of other 66 services. A user can choose to use 66 with runit services and they will still need to have the proper init utilities...
- The '@version' value is set now seperately for each frontend. If a frontend changes than only its '@version' should be incremented. boot@ is of couse an exception.

mobinmob commented:
There is now a template for the first release of void-66-services. It contains 77 services.

travankor commented:
Can we add this to void docs as well?

mobinmob commented:
> Can we add this to void docs as well?

@teldra has written the[ first docs](https://github.com/teldra/void-docs/blob/66-docs/src/config/services/66/index.md) in his void-docs branch. I am revising them since certain things have improved substantially (handling switching poweroff/halt etc transparently and automatically). I will add the doc in this PR and install it with vdoc in the package. I think sending a PR to void-docs is a bit premature at this stage - Ι want to improve certain things first and get rid of [WIP] :)

mobinmob commented:
More work on the runit service:

- Remove the log argument for runsvdir.
- Replace  some utilities (env, mkdir, ln) with their alternatives from s6-portable-utils.
- Change the service` @type` from longrun to classic. That required disabling and re-enabling the service for the current users.

mobinmob commented:
New version of boot-66serv:

- New homepage and distfiles for boot-66serv.
- Needs at least 66 v. 0.6.0.0 - please do not update until https://github.com/void-linux/void-packages/pull/27815 is merged. To use, boot@system should be re-enabled.
- `@options` is removed from services frontend files, as log is always created (unless it is explicitly disabled per-service) and [environment] automaticaly enables `@options = ( env )`.

mobinmob commented:
Rebase onto master and remove tty@ frontend service file - no longer needed, as boot-66serv uses tty-rc@ and tty-earlier@ which are included.

mobinmob commented:
Small improvement:
There is now a symlink in /etc/66rc.conf that points to the configuration file of the `boot@system` service. Now users can edit directly `/etc/66rc.conf` with their favourite editor and then enable again the service with:

`# 66-enable -F -t boot boot@system`

That is strictly to provide a more common approach to editing the basic boot configuration instead of using the - powerfull - `66-env` utility. 
Many thanks to @Obarun for the suggestion of the symlink.

mobinmob commented:
The latest commit adds a small utility script named `66boot-initial-setup`. The utility  automatically creates the recommended trees (boot, default and runit) and enables needed services in them. It is non-interactive. Users can execute it after boot-66serv installation or if they want to start from scratch after removing all their existing trees.

There are also minor changes in the template (double-quoting vars etc).

ericonr commented on srcpkgs/boot-66serv/files/switch-initutils-66:
```diff
@@ -0,0 +1,28 @@
+
```
Don't need the empty line, I think?

mobinmob replied:
Not needed,  just aesthetics :) I can remove it if it is a problem...

ericonr commented on srcpkgs/boot-66serv/files/runit:
```diff
@@ -0,0 +1,31 @@
+@build = custom
+@shebang = ( "/bin/sh" )
+@execute = (
+PATH=/usr/bin:/usr/sbin
```
Why not only `/usr/bin`? And maybe `/usr/local/bin`.

mobinmob replied:
I follow the /etc/runit/2 script. It only enables these these two - one, as /sbin is a symlink to /usr/bin.

mobinmob replied:
Left only /usr/bin. Thank you!

ericonr commented on srcpkgs/boot-66serv/files/66:
```diff
@@ -0,0 +1,2 @@
+#!/usr/bin/execlineb -P
+66-boot -b "Booting Voidlinux" -m
```
VoidLinux or "Void Linux" would be preferred, I think

mobinmob replied:
Void Linux is in the site, so I will change to that.

mobinmob replied:
Done, thank you!

ericonr commented on srcpkgs/boot-66serv/template:
```diff
@@ -0,0 +1,45 @@
+depends="s6 s6-rc s6-linux-utils s6-portable-utils 66 66-tools"
+short_desc="Stage 1 boot for 66"
+maintainer="mobinmob <mobinmob@disroot.org>"
+# previous maintainer="teldra <teldra@rotce.de>"
```
Leave credits only in the commit, IMO.

mobinmob replied:
Done, thank you for the review!

mobinmob commented:
66boot-initial-setup:

- do not overwrite existing /etc/66/conf/boot@system/version/boot@system if it exists.

runit service frontend file:

- redirect stdout to stderr in order to catch everything in the per-service logger,
- use ln instead of s6-ln (different behaviour),
- remove redundant files/services in /var/service.

template:

- disable TMPFILE support by default,
- remove all conf variables from the envronment.

mobinmob commented:
There is now some updated documentation for 66 on voidlinux: https://github.com/mobinmob/void-66-services/blob/devel/conf/void-66-conf.md
@travankor

mobinmob commented:
Changes:
- New void-66-services release. It includes basic documentation (linked in the previous comment). It is not installed with the package yes. Other significant changes are included in the [release changelog](https://github.com/mobinmob/void-66-services/releases/tag/v0.0.2).
- Small fix - typo in the 66boot-initial-setup script.

mobinmob commented:
Changes:

- Remove the warning from the start of the configuration file that is linked in /etc/66rc.conf. That is a tiny usability improvement. The removal happens automatically if the file is generated by using 66-env, but it is just copied in the 66boot-initial-setup script.
- Add two upstream patches. The first fixes a potential  issue discovered by testing this PR with voidvault (encrypted root with btrfs subvolumes). The second makes the iptables service more robust, since it will not need an external script to function.

mobinmob commented:
Changes:

- template : Create the log directory in /var/log/66 with proper permissions and ownership (_s6log). That is not needed for most cases, as it is created automatically by the 66 utilities. The problem is that it can fail - I have seen it happen with zfs datasets mounted under /var/log. Creating it during installation solves this issue.
- files/66boot-initial-setup: use a variable for the new boot@system configuration file and use it consistently.

mobinmob commented:
There is now and unofficial repo to install `boot-66serv` and `void-66services`. The package versions are in the following format:
`upstream.version.number.0.bulddate`
Currently, that is` 2.3.0.0.20210305` for the `boot-66serv `package and `0.0.2.0.20210305` for the `void-66services` package.
The [documentation](https://github.com/mobinmob/void-66-services/blob/devel/conf/void-66-conf.md) has been updated with information on how to enable and use the repo.

mobinmob commented:
Add `scandir-66serv` package. This is just the beggining of enabling proper user services. Documentation and additional scripts will follow.

mobinmob commented:
Changes

template:

- remove the `--ZFS_IMPORT=!zpoolcache` configure switch, as zpoolcache is now the default method.
- remove the creation of the log directory - that is done now by the 66 package.

pathes/
- remove previous patches
- include all changes since the release

paper42 commented:
Files in srcpkgs/boot-66serv/ contain a lot of trailing spaces. Could this PR should be just 3 commits, one for each package?

mobinmob commented:
> Files in srcpkgs/boot-66serv/ contain a lot of trailing spaces. 

Yep, I should clean them up ;)

> Could this PR should be just 3 commits, one for each package?

It should be a commit per-package to get ready for merging. Until there is that possibility -and certain things get improved, I 'd rather have a commit history.

mobinmob commented:
- Rebase on master.
- Remove tmpfiles.d support - it does not exist on voidlinux.

mobinmob commented:
Changes:

 - Update to version 2.4.0.
 - Add back modules.sh (it has moved to a separate repo), add a new
configuration switch for it
 - Rebase the patch to remove tmpfiles, add a fix for complete removal
 - Ρebase all commits to master.

mobinmob commented:
Change:
- Removed `--TMPFILE=!no` from  `configure_args`.  I have removed the support for tmpfiles.d, so the switch is now redundant.

mobinmob commented:
- Rebase on master.
- Fix and rename the tmpfiles.d code removal patch.

New binary packages (20211110 suffix) have been pushed to the [void-66 repo](https://codeberg.org/mobinmob/void-66) ([doc](https://github.com/mobinmob/void-66-services/blob/master/conf/void-66-conf.md)).

thegarlynch commented:
Just one more successful build and i have something to do for weekend

mobinmob commented:
#### Changes for the PR:
- Squashed all commits in oder to have one commit per package. Full history of the changes is retained in the[ 66-voidlinux repo in codeberg ](url) where all new development will take place before updating this PR.
- Rebased on master.

#### Changes for boot-66serv:
- `66boot-initial-setup` is improved significantly.
- Two new configuration scripts are added:
    - `66boot-rcdotconf,` which uses the values of `/etc/rc.conf` to configure boot@system and
    - `66boot-storage-autoconf` which will discover btrfs, zfs, dmraid, cryptsetup volumes and swap partitions/files and configure 66boot-66serv accordingly.
 
Using the three 66boot scripts and re-enabling boot@system should be enough to configure boot@system.

- The upstream `system-sysctl` service now executes the script from runit-void, and the `modules-load` script is used in place of the modules.sh obarun script. Due to these changes` 66-enable -t boot -F boot@system` **needs to run after the update**.

The packages in the void-66 repository will be updated tommorow - the new boot-66serv package with have the version suffix 20220220.

mobinmob commented:
#### boot-66serv changes:

- Install backwards compatibility symlink for removed `/usr/share/66/script/modules.sh` script. That is not needed if someone re-enables boot@system after update but it is a failsafe.
- Add `INSTALL.msg` to remind users to re-enable boot@system after an update.
- Mount efivarfs rw, to have the same behavior as the official runit scheme - reported by @paper42, many thanks :)

thegarlynch commented:
It passed finally, but why didn't it merged yet ? too big ?

mobinmob commented:
> It passed finally, but why didn't it merged yet ? too big ?

Thank you for the interest!
Tests not passing has probably little to do with the possibility of merging this PR :) It essentially adds support for another init system/service manager in the distribution and that is a **big** change. 

The PR  is also the product a totally unofficial effort that goes on for almost two years now and without a promise - or an expectation really - of ever been accepted in the distribution. That is mighty fine for me, as I find the process really rewarding and having enormous fun.

If one needs to use the packages, there is a repo (which will move to osdn at some point), the development for the templates and scripts takes place at the [66-voidlinux](https://codeberg.org/mobinmob/66-voidlinux) repository at codeberg and the service frontend scripts and documentation are maintained in the [void-66-services](https://github.com/mobinmob/void-66-services) repository at github.

mobinmob commented:
## Changes:
- Small fixes for the runit frontend service file (thaks @Obarun).
- A new way to reuse existing runit services - runit-wrapsv@.
- Fix swap detection with 66boot-storage-autoconf.
- Rebase on master.

## Documentation:
- Logging (per-service and system-wide) with 66 - [void-66-logging](https://github.com/mobinmob/void-66-services/blob/master/conf/void-66-logging.md).
- Using runit services with 66 (runit and runit-wrapsv@) - [void-66-runitsv](https://github.com/mobinmob/void-66-services/blob/master/conf/void-66-runitsv.md).

### Pending: 
- Move the packages to the osdn repo - I will anounce it here and update the relevant doc.

mobinmob commented:
## New binary package repo

The temporary binary package repo in codeberg is deprecated.
A new repo is provided, hosted in the [osdn platform]()https://osdn.net/projects/avyssos/.

### Adding the new repository:


_(commands prefixed by `#` must be run with elevated privileges, as root)_
```
# xbps-install -S --repository=https://mirrors.gigenet.com/OSDN//storage/g/a/av/avyssos/repo void-unofficial-repo-66
```

You will be prompted to accept the new key:
```
https://mirrors.gigenet.com/OSDN//storage/g/a/av/avyssos/repo' repository has been RSA signed by "mobinmob"
Fingerprint: c7:39:79:a3:2a:cf:f1:65:a6:df:3a:1a:6e:93:36:28
Do you want to import this public key? [Y/n]  
```

mobinmob commented:
# osdn mirror issue and fix

The mirror that was used in the void-unoficial-repo-66 is not working, so the user will not be able to get future updates. A new repo package is uplaoded in codeberg.org/mobinmob/66-void that fixes that issue. In order to install use the following command:
`
# xbps-install --repository=https://codeberg.org/mobinmob/void-66/raw/branch/master  void-unofficial-repo-66`

The packages are signed with the same key.

NymanMatthias commented:
With the removal of tmpfiles.d support, I believe that rc-local should no longer depend on local-tmpfiles, as it doesn't exist anymore.

mobinmob commented:
> With the removal of tmpfiles.d support, I believe that local-rc should no longer depend on local-tmpfiles, as it doesn't exist anymore.

That is correct, thank you for the report!
I will fix it in the next package upload and report here.
I have some unfinished work for the upstream boot-66serv that I would like to finish first, as it is making the boot process more robust : https://framagit.org/pkg/obmods/boot-66serv/-/merge_requests/17

mobinmob commented:
> I have some unfinished work for the upstream boot-66serv that I would like to finish first, as it is making the boot process more robust : https://framagit.org/pkg/obmods/boot-66serv/-/merge_requests/17

Work continues in the obarun framagit instance:
https://git.obarun.org/obmods/boot-66serv/-/merge_requests/1#54776d84710e7228457fa1df3eecc1319a5e57fe

I am rebasing the voidlinux patches on top of this branch, an update is near :)

mobinmob commented:
## Changes

#### boot-66serv

- Μany checks added to the configure script that runs when the service is enabled:
	- check for existance of keys and the validity of their values
	- discover luks,lvm,zfs,btrfs.dmraid,mdraid and check for the correct configuration of the relecant keys
	- add mdraid support
	- add a staged approach to the `devices-*` services in order to avoid breaking boot and inform the user for proper configuration
	- add a configuration key to enable/disable the checks (CHECK_CONFIGURATION=yes by default)
	- ammend the documentation for the changes.
	*The changes above are explained [here](https://git.obarun.org/obmods/boot-66serv/-/merge_requests/1#note_121)*
- Remove the tmpfiles dependency from local-rc (thanks @NymanMatthias for the report).
- Fix zfs detection on `66boot-storage-autoconf`.
- Improve `switch-initutils-66` frontend service file.
- Change of license to 0BSD (upstream change, the `66boot-*` utilities remain under BSD-2-Clause).
- Update to 2.4.1

#### 66-void and base-system-66 (new packages)

66-void can replace runit-void and base-system-66 can replaces base-system, for a more "clean" and correct approach to changing init systems in voidlinux.

#### void-66-services

The package is not updates, but many service frontend files have beed added to the repo since the last release and there are also significant fixes.
Everyone that needs them is encouraged to use the void-66-services-master template under packaging/ in the github repo.
I want to finish and test the user/session services work and release 0.0.3 with it.

#### Documentation 
Document the new packages and the transition to/from them:
[void-66-services/void-66-base-system.md at master · mobinmob/void-66-services · GitHub](https://github.com/mobinmob/void-66-services/blob/master/conf/void-66-base-system.md)

*Packages are uploaded to the repo, the package version for boot-66serv is 2.4.1.20221016.*

NymanMatthias commented:
I spun up a VM to test this PR, and everything seems to work fine. One thing I noticed while setting it up is that some of these packages are missing dependencies on s6-rc, s6, and execline. Without these packages installed, boot-66serv will never bring Void to a functional state. On Obarun, these are all dependencies of the 66 package, while Void lacks these dependencies.

mobinmob commented:
> I spun up a VM to test this PR, and everything seems to work fine. One thing I noticed while setting it up is that some of these packages are missing dependencies on s6-rc, s6, and execline. Without these packages installed, boot-66serv will never bring Void to a functional state. On Obarun, these are all dependencies of the 66 package, while Void lacks these dependencies.

Thank you for testing!
The 66 package should depend on these implicitly, since its template contains  `makedepends="oblibs-devel skalibs-devel execline-devel s6-devel s6-rc-devel"`. The other packages depend on it. I will check the installation again and see why there is an issue. OFC 66 cannot work without s6 and s6-rc :)

mobinmob commented:
@NymanMatthias : https://github.com/void-linux/void-packages/pull/40007
s6 and execline are s6-rc deps. Thanks for the report :)

mobinmob commented:
Ι am currently testing 66 with the new skarnet stack in https://github.com/void-linux/void-packages/pull/41635 .  After that I will push a small update (fixes) here and in the repo.

mobinmob commented:
# Changes

## boot-66ser:
- Fixes for zfs detection in _66boot-storage-autoconf_
- Small changes in 66boot-* scripts.

## void-66services v. 0.0.3 :

- 181 services in total!
- More work to have services log to the per-service logger instead of syslog.
- Add a changelog.
- Split licensing information (LICENSE - LICENSE-3RD-PARTY) and amend the README.md accordingly.
- Repo-wide: remove unneeded @options key, standardise layout a little.
- Create additional service frontend files for nginx, dhcpcd and dhcpcd@ that use pid namespaces though 66-ns. These frontends have the suffix _ns after the service name.
- All services are represented now by either a single frontend file or a single directory with necessary files/subdirs.
- Use a single template that packages the master branch for the users that want to track development - void-66-services-master.
- Include documentantion in the package, starting from void-66-services-master and version 0.0.3.
- Documentation:
  - Add simple logging documentation in conf/void-66-logging.md.
  - Add information about the new 66boot-* utilities in conf/void-66-conf.md
  - Change Suggestions_for_services.md to point to the master branch as the development branch.
  - Add documentation for running runit services with the runit and runit-wrapsv@ service frontend files in conf/void-66-runitsv.md. These service frontend files are not in this repo, but in the [66-voidlinux](https://codeberg.org/mobinmob/66-voidlinux) repo since they are currently part of the boot-66serv package.
  - Change the instructions in conf/void-66-conf.md to use the new binary packages repo.
  - Add documentation for the new 66-void and base-system-66 packages in conf/void-66-base-system.md.
- Many fixes.
- Two new contributors in this release, thanks @linuxergr and @paper42 

Packages with be appearing in the repo in an hour or so with the suffix (when applicable) `20230205`.
The [66-voidlinux repo](https://codeberg.org/mobinmob/66-voidlinux/src/branch/master/srcpkgs) with have the same tag.

mobinmob commented:
###Changes:
 - boot-66serv : replaces removed 66-test with test
 - 66-void: new version, remove `/etc/os-release`, it is now in `base-files`.
 
Packages with be appearing in the repo in  less than a day with the version suffix (when applicable) 20230311.
The [66-voidlinux repo](https://codeberg.org/mobinmob/66-voidlinux/src/branch/master/srcpkgs) with have the same tag.

sonnysighedup commented:
Hi @mobinmob, 

I get an interesting crash after the recent 66/s6/execline/oblibs/skalibs update you pushed to the void repos recently. The update itself went fine, but when I go to reboot, the system crashes, and upon hard reboot - after nearly making it all the way through the boot - I get the following message:

`if: fatal: unable to spawn s6-test: No such file or directory`
`if: fatal: unable to spawn s6-test: No such file or directory`
`ifelse: fatal: unable to spawn s6-test: No such file or directory`
`error: i/o error: Broken pipe (os error 32)`

I was able to downgrade everything and boot as normal.

I think the crash happened right before greetd was to pull up my login page. Could this be related to the shift from s6-test to test in boot-66serv from last month?

mobinmob commented:
> Hi @mobinmob, 
> 
> I get an interesting crash after the recent 66/s6/execline/oblibs/skalibs update you pushed to the void repos recently. The update itself went fine, but when I go to reboot, the system crashes, and upon hard reboot - after nearly making it all the way through the boot - I get the following message:
> 
> `if: fatal: unable to spawn s6-test: No such file or directory`
> `if: fatal: unable to spawn s6-test: No such file or directory`
> `ifelse: fatal: unable to spawn s6-test: No such file or directory`
> `error: i/o error: Broken pipe (os error 32)`
> 
> I was able to downgrade everything and boot as normal.
> 
> I think the crash happened right before greetd was to pull up my login page. Could this be related to the shift from s6-test to test in boot-66serv from last month?

Since the last update of this PR and the repo packages in avyssos I have removed the single reference to s6-test from boot-66serv. [I also pushed the one-line change upstream and was acccepted and merged.](https://git.obarun.org/obmods/boot-66serv/-/merge_requests/2).
Are you using the packages from the repo? The latest boot-66serv version is

sonnysighedup commented:
Thank you for the quick reply, @mobinmob.

It's turns out this was more of a display manager issue, than a boot one. I forgot that I hacked up [ob-gtkgreet-sway](https://framagit.org/xprl-obarun-contrib-wayland/ob-gtkgreet-sway) to work with your void-66 implementation. The `wayland-session` script used to launch window managers calls `s6-test`. I replaced those references with `test` and everything worked with the updated 66/s6 packages, though it produces some output that I haven't been able to catch yet.

Thanks for all work you've to make 66 a reality on void!

mobinmob commented:
Thank you for the feedback - always happy to help 😉
If you have working frontend files for your setup which are missing from void-66-services, please consider sending s PR.

mobinmob commented:
osnd issues means the unofficial repo is down - I will have a solution hopefully in the nect week.

mobinmob commented:
@paper42 Can this PR be re-opened?

mobinmob commented:
New repo is up, please add it with:

`# xbps-install -S --repository=https://codeberg.org/mobinmob/void-66/raw/branch/master void-unofficial-repo-66`

or

`# xbps-install -S --repository=http://www.avyssos.eu/repos/voidlinux/66 void-unofficial-repo-66`

The new release tar is `20230805`, has only minor xlint fixes. I pushed it to my branch but since this PR is still closed it will not a ppear here...

mobinmob commented:
The repo now supports tls, the [void-unofficial-repo-66 package was updated](https://codeberg.org/mobinmob/66-voidlinux/commit/721050e92677a728f2c4f4eccf8b8969eb21447d) to version 4 to reflect that. No user intervention is required other than a regular update :)

mobinmob commented:
The PR continues in https://github.com/void-linux/void-packages/pull/45578.
Thanks for eveyone that contributed, I hope you are still interested in 66 on voidlinux. Many nice develpments lie ahead :P
